pkg/knode-manager/controllers/pv.go

```diff
@@ -0,0 +1,298 @@
+		return
+	}
+	if pvInstanceDeepCopy.Spec.ClaimRef != nil || upstreamPV.Spec.ClaimRef == nil {
+		claim := pvInstanceDeepCopy.Spec.ClaimRef
+		var newPVC *corev1.PersistentVolumeClaim
+		newPVC, err = p.client.CoreV1().PersistentVolumeClaims(claim.Namespace).Get(context.TODO(), claim.Name, metav1.GetOptions{})
+		if err != nil {
+			klog.Errorf("kosmos sync downstream: get pvc from upstream failed, pv: %v", pvInstance.Name)
+			return
+		}
+		pvInstanceDeepCopy.Spec.ClaimRef.UID = newPVC.UID
+		pvInstanceDeepCopy.Spec.ClaimRef.ResourceVersion = newPVC.ResourceVersion
+	}
+
+	_, err = p.patchPV(context.TODO(), p.client, upstreamPV, pvInstanceDeepCopy)
+	if err != nil {
+		klog.Errorf("kosmos sync downstream: patch pv from upstream failed, pv: %v", pvInstance.Name)
+		return
+	}
+	p.downstreamQueue.Forget(key2pv)
+	klog.Infof("kosmos sync downstream: finish sync downstream pv, pv: %v", pv2key.(string))
+}
+
+func (p *PVController) deletePV(obj interface{}) {
+	delPV := obj.(*corev1.PersistentVolume)
+	if delPV.Namespace == metav1.NamespaceSystem || delPV.Name == "kubernetes" {
+		key, err := cache.MetaNamespaceKeyFunc(p)
+		if err != nil {
+			runtime.HandleError(err)
+			return
+		}
+		p.upstreamQueue.Add(key)
+		klog.Infof("kosmos delete pv, enqueue: %v", key)
+	} else {
+		klog.Infof("kosmos ignore pv change, pv: ", delPV.Name)
+	}
+}
+
+func (p *PVController) patchPV(ctx context.Context, c kubernetes.Interface, old, new *corev1.PersistentVolume) (*corev1.PersistentVolume, error) {
+	if reflect.DeepEqual(old.Annotations, new.Annotations) && reflect.DeepEqual(old.Spec, new.Spec) && reflect.DeepEqual(old.Status, new.Status) {
+		return old, nil
+	}
+
+	new.Spec.NodeAffinity = old.Spec.NodeAffinity
+	new.UID = old.UID
+	new.ResourceVersion = old.ResourceVersion
+	oldByte, err := json.Marshal(old)
+	if err != nil {
+		return nil, err
+	}
+	newByte, err := json.Marshal(new)
+	if err != nil {
+		return nil, err
+	}
+	patchByte, err := jsonpatch.CreateMergePatch(oldByte, newByte)
+	if err != nil {
+		return nil, err
+	}
+
+	newPV, err := c.CoreV1().PersistentVolumes().Patch(ctx, old.Name, types.MergePatchType, patchByte, metav1.PatchOptions{})
+	if err != nil {
+		return old, err
+	}
+
+	return newPV, nil
+}
+
+func (p *PVController) filterPV(pv *corev1.PersistentVolume, factor string) error {
+	utils.TrimObjectMeta(&pv.ObjectMeta)
+	if pv.Spec.NodeAffinity == nil {
+		return fmt.Errorf("kosmos filter pv failed, pv: %v", pv.Name)
+	}
+	if pv.Spec.NodeAffinity.Required == nil {
+		return fmt.Errorf("kosmos filter pv failed, pv: %v", pv.Name)
+	}
+
+	nsTerms := pv.Spec.NodeAffinity.Required.NodeSelectorTerms
+	for k, v := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		mFields := v.MatchFields
+		mExpressions := v.MatchExpressions
+		for k1, v1 := range v.MatchFields {
+			if v1.Key == "kubernetes.io/hostname" || v1.Key == "beta.kubernetes.io/hostname" {
+				v1.Values = []string{factor}
```

Comment:
you can find "kubernetes.io/hostname" const in ./pkg/knode-manager/utils/constants.go

Reply:
ok, i'm done